adam/perception/perception_graph.py

```diff
@@ -2932,10 +2946,168 @@ def _translate_frames(
+                    action_stroke_graph["joint_points_world_coords"],
+                    action_stroke_graph["joint_points_xyd_coords"],
+                    action_stroke_graph["joint_points_confidence_score"],
+                )
+            ):
+                world_coord, xyd_coord, conf = joint_feature
+                joint_point_nodes.append(
+                    JointPointNode(
+                        scene_xyd_coord=DepthPoint(
+                            xyd_coord[0], xyd_coord[1], xyd_coord[2]
+                        ),
+                        world_coord=Point(world_coord[0], world_coord[1], world_coord[2]),
+                        confidence=conf,
+                        weight=conf,
+                        temporal_index=i,
+                    )
+                )
+
+            # First we link the objects to the action node
+            edges_to_add: List[Tuple[GraphNode, GraphNode, OntologyNode]] = [
+                (object_node, action_decode_node, INVOLVED_IN_ACTION)
+                for object_node in involved_object_nodes
+            ]
+            edges_to_add.extend(
+                (action_decode_node, joint_node, JOINT_POINT_TRAVERSAL)
+                for joint_node in joint_point_nodes
+            )
+            edges_to_add.extend(
+                (agent_node, joint_node, HAS_PROPERTY_LABEL)
+                for joint_node in joint_point_nodes[1:-1]
+            )
+
+            # Special Case the first and last joint point as 'before' and 'after'
+            # This probably isn't needed but it's closer to temporally correct
+            _dynamic_digraph.add_edge(
+                agent_node,
+                joint_point_nodes[0],
+                label=TemporallyScopedEdgeLabel.for_dynamic_perception(
+                    HAS_PROPERTY_LABEL, TemporalScope.BEFORE
+                ),
+            )
+            _dynamic_digraph.add_edge(
+                agent_node,
+                joint_point_nodes[-1],
+                label=TemporallyScopedEdgeLabel.for_dynamic_perception(
+                    HAS_PROPERTY_LABEL, TemporalScope.AFTER
+                ),
+            )
+
+            # Also to help clarify the agent we mark them as 'VOLITIONAL' from the original symbolic system
+            # This is mostly a way of representing the structured knowledge in the feature file.
+            _dynamic_digraph.add_edge(
+                agent_node,
+                VOLITIONALLY_INVOLVED,
+                label=TemporallyScopedEdgeLabel.for_dynamic_perception(
+                    HAS_PROPERTY_LABEL, [TemporalScope.BEFORE, TemporalScope.AFTER]
+                ),
+            )
+
+            for source, target, label in edges_to_add:
+                _dynamic_digraph.add_edge(
+                    source,
+                    target,
+                    label=TemporallyScopedEdgeLabel.for_dynamic_perception(
+                        label, TemporalScope.DURING
+                    ),
+                )
```

Comment:
(:banana:) It would be nice to break the different categories of edges into smaller helper functions. Say, for example, `_link_action_to_objects()`, `_link_action_to_agent()`, and `_link_agent_to_joint_points()`.